include/zephyr/drivers/sensor.h

```diff
@@ -394,12 +398,171 @@ typedef int (*sensor_channel_get_t)(const struct device *dev,
 				    enum sensor_channel chan,
 				    struct sensor_value *val);
 
+/**
+ * @typedef sensor_frame_iterator_t
+ * @brief Used for iterating over the data frames via the :c:struct:`sensor_decoder_api`
+ *
+ * Example usage:
+ *
+ * @code(.c)
+ *     sensor_frame_iterator_t fit = {0}, fit_last;
+ *     sensor_channel_iterator_t cit = {0}, cit_last;
+ *
+ *     while (true) {
+ *       int num_decoded_channels;
+ *       enum sensor_channel channel;
+ *       q31_t value;
+ *
+ *       fit_last = fit;
+ *       num_decoded_channels = decoder->decode(buffer, &fit, &cit, &channel, &value, 1);
+ *
+ *       if (num_decoded_channels <= 0) {
+ *         printk("Done decoding buffer\n");
+ *         break;
+ *       }
+ *
+ *       printk("Decoded channel (%d) with value %s0.%06" PRIi64 "\n", q < 0 ? "-" : "",
+ *              abs(q) * INT64_C(1000000) / (INT64_C(1) << 31));
+ *
+ *       if (fit_last != fit) {
+ *         printk("Finished decoding frame\n");
+ *       }
+ *     }
+ * @endcode
+ */
+typedef uint32_t sensor_frame_iterator_t;
+
+/**
+ * @typedef sensor_channel_iterator_t
+ * @brief Used for iterating over data channels in the same frame via :c:struct:`sensor_decoder_api`
+ */
+typedef uint32_t sensor_channel_iterator_t;
+
+/**
+ * @brief Decodes a single raw data buffer
+ *
+ * Data buffers are provided on the :c:struct:`rtio` context that's supplied to
+ * c:func:`sensor_read`.
+ */
+struct sensor_decoder_api {
+	/**
+	 * @brief Get the number of frames in the current buffer.
+	 *
+	 * @param[in]  buffer The buffer provided on the :c:struct:`rtio` context.
+	 * @param[out] frame_count The number of frames on the buffer (at least 1)
+	 * @return 0 on success
+	 * @return <0 on error
+	 */
+	int (*get_frame_count)(const uint8_t *buffer, uint16_t *frame_count);
+
+	/**
+	 * @brief Get the timestamp associated with the first frame.
+	 *
+	 * @param[in]  buffer The buffer provided on the :c:struct:`rtio` context.
+	 * @param[out] timestamp_ns The closest timestamp for when the first frame was generated as
+	 *             attained by :c:func:`k_uptime_get`.
+	 * @return 0 on success
+	 * @return <0 on error
+	 */
+	int (*get_timestamp)(const uint8_t *buffer, uint64_t *timestamp_ns);
```

Comment:
@teburd @MaureenHelm I didn't think that ms resolution was good enough but I understand that many boards will not have ns resolution either.
Design choice:
- [ ] Use microseconds
- [X] Use nanoseconds
- [ ] Use ticks

I went with nanoseconds, but honestly I'm leaning towards just ticks and letting the user compute it since there are plenty of helpers in kernel.h

Reply:
Updated to ticks